lib/AST/ASTScopeCreation.cpp

```diff
@@ -1609,6 +1604,12 @@ ASTScopeImpl *GenericTypeOrExtensionWholePortion::expandScope(
       scope->getDecl(), genericParams, scope);
   if (context->getTrailingWhereClause())
     scope->createTrailingWhereClauseScope(deepestScope, scopeCreator);
+
+  // Prevent circular request bugs caused by illegal input and
+  // doing lookups that getExtendedNominal in the midst of getExtendedNominal.
+  if (scope->shouldHaveABody() && !scope->doesDeclHaveABody())
```

Comment:
Can this happen in non-code completion cases too?

Reply:
Actually, yes. For example, for:
```swift
extension Collection where Element == Int
```
current master emits 
```
test.swift:2:42: error: expected '{' in extension
extension Collection where Element == Int
                                         ^
test.swift:2:28: error: cannot find type 'Element' in scope
extension Collection where Element == Int
                           ^~~~~~~
test.swift:2:36: error: generic signature requires types '<<error type>>' and 'Int' to be the same
extension Collection where Element == Int
                                   ^
```
With this change, it's just:
```
test.swift:2:42: error: expected '{' in extension
extension Collection where Element == Int
                                         ^
```

Reply:
Which file can I add a test case for this?

Reply:
Added to `test/decl/ext/extensions.swift`